race-buffer/src/reader.rs

```diff
@@ -0,0 +1,139 @@
+use crate::{get_cursor_index, Entry};
+use core::marker::PhantomData;
+use std::cmp::min;
+use std::error::Error;
+
+type Result<T> = std::result::Result<T, Box<dyn Error>>;
+/// Used to "snap" observations of the RaceBuffer's write cursor
+/// or entries in the RaceBuffer's backing storage. Each Snapper access
+/// should have sequential consistency with the memory access of the RaceBuffer writer
+pub trait Snapper<E>
+where
+    E: Entry,
+{
+    /// Take a snapshot of the write cursor of the RaceBuffer
+    fn snap_wcurs(&self) -> Result<usize>;
+    /// Take a snapshot of the overwrite cursor of the RaceBuffer
+    fn snap_owcurs(&self) -> Result<usize>;
+    /// Take a snapshot of the RaceBuffer's backing storage at the given index
+    fn snap_storage(&self, index: usize) -> Result<E>;
+}
+
+/// Struct used to read from a RaceBuffer asynchronously
+pub struct RaceBufferReader<E, S>
+where
+    E: Entry,
+    S: Snapper<E>,
+{
+    /// Struct for reading writer state
+    snapper: S,
+    /// Reader's cursor
+    rcurs: usize,
+    /// Capacity of backing storage
+    storage_cap: usize,
+    /// Cached prefix, not put into buffer until suffix is successfully read
+    stored_prefix: Option<E>,
+    phantom: PhantomData<E>,
+}
+
+impl<E, S> RaceBufferReader<E, S>
+where
+    E: Copy + PartialEq + Entry,
+    S: Snapper<E>,
+{
+    /// Create new RaceBufferReader
+    pub fn new(snapper: S, storage_cap: usize) -> RaceBufferReader<E, S> {
+        RaceBufferReader {
+            snapper,
+            rcurs: 0,
+            storage_cap,
+            stored_prefix: None,
+            phantom: PhantomData,
+        }
+    }
+
+    /// Attempt to read all new entries in race buffer into given read buffer
+    pub fn read(&mut self, rbuf: &mut Vec<Option<E>>) -> Result<()> {
+        let pre_wcurs = self.snapper.snap_wcurs()?;
+        let pre_owcurs = self.snapper.snap_owcurs()?;
+
+        if pre_owcurs > self.rcurs + self.storage_cap {
+            // If writer has overwritten unread entries between reads, store nils and correct rcurs
+            let num_missed = pre_owcurs - (self.rcurs + self.storage_cap);
+            self.store_missed_items(num_missed, rbuf);
+            self.rcurs = pre_owcurs - self.storage_cap;
+        }
+
+        // Perform reads into snapshot buffer up to write cursor
+        let first_read = self.rcurs;
+        let mut buf_snapshot: Vec<E> = Vec::new();
```

Comment:
I'm unclear on why we need this intermediate buffer. Can we write directly to `rbuf` (and manage any necessary initial-index-offsetting needed)?

Reply:
This is possible, but it would require copying the entries anyway if any items were missed during the read.

Reply:
Reducing the number of involved entities is probably a general win for code understandability. Removing the intermediate Vec also cuts out some allocation, too.  Code clarity is the higher priority, and since it's pretty subjective, so I wouldn't call this quibble essential.